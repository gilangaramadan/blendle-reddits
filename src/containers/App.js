import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { fetchPosts } from "../store/actions";

import Posts from "../components/Posts";

class App extends Component {
  constructor(props) {
    super(props);
    this.shouldComponentRender = this.shouldComponentRender.bind(this);
  }

  componentDidMount() {
    const { fetchPosts } = this.props;
    fetchPosts();
  }

  shouldComponentRender() {
    const { pending } = this.props;
    if (pending === false) return false;
    // more tests
    return true;
  }

  render() {
    const { posts, error, pending } = this.props;
    return (
      <React.Fragment>
        <div className="header home">
          <h1>Home</h1>
          <small>Top 10 posts</small>
        </div>
        <div style={{ opacity: pending ? 0.5 : 1 }}>
          {this.shouldComponentRender() && posts.length === 0 && (
            <span className="text-helper">Loading...</span>
          )}
          {!this.shouldComponentRender() && posts.length === 0 && (
            <span className="text-helper">Empty.</span>
          )}
          {posts.length > 0 && <Posts posts={posts} />}
          {error && <span className="text-helper error">{error}</span>}
        </div>
      </React.Fragment>
    );
  }
}

App.propTypes = {
  posts: PropTypes.array.isRequired,
  pending: PropTypes.bool.isRequired,
  error: PropTypes.object
};

const mapStateToProps = state => ({
  error: state.error,
  posts: state.posts,
  pending: state.pending
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      fetchPosts
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);

export { App };
