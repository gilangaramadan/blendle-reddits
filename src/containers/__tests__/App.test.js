import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { App } from "../App";

describe("App", () => {
  it("renders without crashing given the required props", () => {
    const props = {
      error: null,
      posts: [],
      pending: false,
      fetchPosts: jest.fn()
    };
    const wrapper = shallow(<App {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
