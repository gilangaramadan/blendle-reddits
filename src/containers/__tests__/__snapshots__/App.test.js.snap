// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing given the required props 1`] = `
<Fragment>
  <div
    className="header home"
  >
    <h1>
      Home
    </h1>
    <small>
      Top 10 posts
    </small>
  </div>
  <div
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <span
      className="text-helper"
    >
      Empty.
    </span>
  </div>
</Fragment>
`;
